docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,960 @@
+  whose data is being deleted are current.
+
+The order of operations for correctness is:
+
+1. Decide to delete one or more objects
+2. Write updated metadata (this means that if we are still the rightful leader, any subsequent leader will
+   read metadata that does not reference the about-to-be-deleted object).
+3. Confirm we hold a non-stale generation for the node and the attachment(s) doing deletion.
+4. Actually delete the objects
+
+Note that at stage 3 we are only confirming that deletions of objects no longer referenced
+by the metadata written in step 2 are safe. While Step 3 is in flight, new deletions may
+be enqueued, but these are not safe to execute until a subsequent iteration of the
+generation check.
+
+Because step 3 puts load on the control plane, deletion lists should be executed lazily,
+accumulating many deletions before executing them behind a batched generation check message
+to the control plane, which would contain a list of tenant generation numbers to validate.
+
+The result of the request to the control plane may indicate that only some of the attachment
+generations are fresh: if this is the case, then the pageserver must selectively drop the
+deletions from a stale generation, but still execute the deletions for attachments
+that had a fresh generation.
+
+The only correctness requirement for deletions is that they are _not_ executed prior
+to the barrier in step 3. It is safe to delay a deletion, or indeed to never execute
+it at all, if a node restarts while an in-memory queue of deletions is pending. Minimizing
+leaking objects is an optimization, accomplished by making this queue persistent: see
+[Persistent Deletion Queue](#persistent-deletion-queue) in the optimizations section.
+
+### Deletion Part 2: WAL trim changes (delay `remote_consistent_lsn` updates)
+
+Remote objects are not the only kind of deletion the pageserver does: it also indirectly deletes
+WAL data, by feeding back remote_consistent_lsn to safekeepers, as a hint to the safekeepers that
+they may drop data below this LSN.
+
+We may solve for safety of remote_consistent_lsn updates in the same way as for deletions: before indicating to safekeepers
+that they may trim up to a particular LSN, pageservers should communicate with the control plane
+to confirm that their node generation is current. This guarantees that any future generation will
+be starting from the last metadata that this generation wrote out, i.e. the S3 state which contains
+all the data from the WAL up to the proposed trim point.
+
+The pageservers indicate what to trim by the `remote_consistent_lsn` field in messages to
+the safekeeper. The pageservers will change to only advertise the true remote_consistent_lsn
+after they have periodically checked the validity of their generation: this will slightly
+delay WAL trimming by whatever period we decide to make that control plane check-in.
```

Comment:
The first paragraph is too fuzzy wrt "generation number" for my taste.
You only mention "node generation number".
What about "attachment generation number"?

Do we need to do a check-in for both? If so, what does it mean if one is recent but the other is not?

Given your recent changes to this section, i suggest to remove all but the first sentence of the first paragraph.

But even after that, there remains a fuzzy "their generation". 

Please clarify further.

```suggestion
We may solve for safety of remote_consistent_lsn updates in the same way as for deletions:
the pageservers indicate what to trim by the `remote_consistent_lsn` field in messages to
the safekeeper. The pageservers will change to only advertise the true remote_consistent_lsn
after they have periodically checked the validity of their generation: this will slightly
delay WAL trimming by whatever period we decide to make that control plane check-in.
```

Reply:
Reworked this section to be shorter in 21bf2e6320f62f3404805b1af0d542c07c3bc681